lib/parallel.c

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright The rpminspect Project Authors
+ * SPDX-License-Identifier: LGPL-3.0-or-later
+ */
+#include <unistd.h>
+#include <stdlib.h>
+#include <inttypes.h>
+#include <errno.h>
+#include <err.h>
+#include <sched.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <stdio.h>
+#include <string.h>
+#include "rpminspect.h"
+#include "parallel.h"
+
+unsigned default_parallel_processes = 0;
+
+static unsigned available_cpus(void)
+{
+#if 0
+    /* This reads /sys/devices/system/cpu/online, which isn't affected by CPU mask */
+    return sysconf(_SC_NPROCESSORS_ONLN);
+#else
+    /* We want "taskset 0x7 rpminspect ..." to correctly assume that only 3 CPUs are available */
+    int r;
+    uint32_t mask[4096 / 4];
+
+    memset(mask, 0, sizeof(mask));
+    r = sched_getaffinity(0, sizeof(mask), (void*)mask);
+    if (r != 0) {
+        errx(EXIT_FAILURE, "I am not prepared for machines with 4500+ CPUs");
+        /* we can write a (re)allocating/looping version when we need to */
+    }
+    r = 0;
+    for (unsigned i = 0; i < sizeof(mask)/sizeof(mask[0]); i++) {
+        uint32_t m = mask[i];
+        if (m == 0) continue;
+        if (~m == 0) { /* fully all-ones word (typical) */
+            r += sizeof(mask[0]) * 8;
+            continue;
+        }
+        /* obscure method to count bits in 32-bit word */
+        m = m - ((m >> 1) & 0x55555555);
+        m = (m & 0x33333333) + ((m >> 2) & 0x33333333);
+        r += (((m + (m >> 4)) & 0x0f0f0f0f) * 0x01010101) >> 24;
```

Comment:
Someone is going to come along a read this later and not know what to think.  Whenever I have stuff like this in code, I prefer to have a comment block explaining what it's doing for the reader beyond just a description.

Reply:
I didn't investigate it either, I just took it from other code (and tested that it works).

Reply:
Found this rather terse explanation:
    uint32_t i = v;
     i = i - ((i >> 1) & 0x55555555);        // add pairs of bits
     i = (i & 0x33333333) + ((i >> 2) & 0x33333333);  // quads
     i = (i + (i >> 4)) & 0x0F0F0F0F;        // groups of 8
     i *= 0x01010101;                        // horizontal sum of bytes in top byte
     i = i >> 24;               // isolate just top byte

Reply:
How this SWAR bithack works:

i = i - ((i >> 1) & 0x55555555);

The first step is an optimized version of masking to isolate the odd / even bits, shifting to line them up, and adding. This effectively does 16 separate additions in 2-bit accumulators (SWAR = SIMD Within A Register). Like (i & 0x55555555) + ((i>>1) & 0x55555555).

The next step takes the odd/even eight of those 16x 2-bit accumulators and adds again, producing 8x 4-bit sums. The i - ... optimization isn't possible this time so it does just mask before / after shifting. Using the same 0x33... constant both times instead of 0xccc... before shifting is a good thing when compiling for ISAs that need to construct 32-bit constants in registers separately.

The final shift-and-add step of (i + (i >> 4)) & 0x0F0F0F0F widens to 4x 8-bit accumulators. It masks after adding instead of before, because the maximum value in any 4-bit accumulator is 4, if all 4 bits of the corresponding input bits were set. 4+4 = 8 which still fits in 4 bits, so carry between nibble elements is impossible in i + (i >> 4).

So far this is just fairly normal SIMD using SWAR techniques with a few clever optimizations. Continuing on with the same pattern for 2 more steps can widen to 2x 16-bit then 1x 32-bit counts. But there is a more efficient way on machines with fast hardware multiply:

Once we have few enough "elements", a multiply with a magic constant can sum all the elements into the top element. In this case byte elements. Multiply is done by left-shifting and adding, so a multiply of x * 0x01010101 results in x + (x<<8) + (x<<16) + (x<<24). Our 8-bit elements are wide enough (and holding small enough counts) that this doesn't produce carry into that top 8 bits.